Sources/PackageRegistry/SourceArchiver.swift

```diff
@@ -0,0 +1,72 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2021 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import Basics
+import TSCBasic
+import TSCUtility
+
+import Dispatch
+
+/// An `Archiver` that handles source archives.
+///
+/// Source archives created with `swift package archive-source`
+/// have a top-level directory prefix (for example, `LinkedList-1.1.1/`).
+/// Unfortunately, the `unzip` command used by `ZipArchiver` doesn't have an option for
+/// ignoring this top-level prefix.
+/// Rather than performing additional (possibly unsafe) file system operations,
+/// this `Archiver` delegates to `tar`, which has a built-in `--strip-components=` option.
+struct SourceArchiver: Archiver {
+    public var supportedExtensions: Set<String> { ["zip"] }
+
+    /// The file-system implementation used for various file-system operations and checks.
+    private let fileSystem: FileSystem
+
+    /// Creates a `SourceArchiver`.
+    ///
+    /// - Parameters:
+    ///   - fileSystem: The file-system to used by the `SourceArchiver`.
+    public init(fileSystem: FileSystem = localFileSystem) {
+        self.fileSystem = fileSystem
+    }
+
+    public func extract(
```

Comment:
I think there is some code in workspace that also does similar things (for binary artifacts). maybe we can reuse / converge on one implementation?

Reply:
There's `ZipArchiver` in TSCUtility, which I based this type on: https://github.com/apple/swift-tools-support-core/blob/f9bbd6b80d67408021576adf6247e17c2e957d92/Sources/TSCUtility/Archiver.swift#L34

(Hence some of the callouts from copy-pasted code, like dispatching to the global queue)

I discuss the rationale for this type in the doc comments, but the main takeaway is that this is probably only useful for the registry. I'd also be interested in getting some more 👀 on this, especially from security folks.  

As far as converging on a single implementation, there may be an opportunity to make the delegated system call and directory prefix stripping a configurable property. But at that point, I think you may as well have two separate implementations of the `Archiver` protocol.

Reply:
cc @andyp-apple and @robertlacroix